### PR TITLE
Add tests for usage of IsArray

### PR DESCRIPTION
**Pull request by jugglinmike**
Interestingly, SpiderMonkey fails just one of these specific cases. The
following code is expected to throw a TypeError, but does not in SpiderMonkey:

    var h = Proxy.revocable({}, {});
    h.revoke();
    Object.prototype.toString.call(h.proxy);

### CODE REVIEW

File: test/built-ins/Array/isArray/proxy.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+---*/
+
+var objectProxy = new Proxy({}, {});
+var arrayProxy = new Proxy([], {});
+var arrayProxyProxy = new Proxy(arrayProxy, {});
+
+assert.sameValue(Array.isArray(objectProxy), false);
+assert.sameValue(Array.isArray(arrayProxy), true, 'proxy for array');
+assert.sameValue(
+  Array.isArray(arrayProxyProxy), true, 'proxy for proxy for array'
```
Comment (leobalter) [inline test/built-ins/Array/isArray/proxy.js]: lol, the name is funny and it's great IsArray handling proxy targets.

File: test/built-ins/Array/prototype/concat/create-proxy.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+var proxy = new Proxy(new Proxy(array, {}), {});
+var Ctor = function() {};
+var result;
+
+array.constructor = function() {};
+array.constructor[Symbol.species] = Ctor;
+
+result = Array.prototype.concat.call(proxy);
+
+assert.sameValue(Object.getPrototypeOf(result), Ctor.prototype);
```
Comment (leobalter) [inline test/built-ins/Array/prototype/concat/create-proxy.js]: nice!

File: test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+     c. Return ? IsArray(target).
+features: [Proxy]
+---*/
+
+var handle = Proxy.revocable([], {});
+
+handle.revoke();
+
+assert.throws(TypeError, function() {
+  [].concat(handle.proxy);
```
Comment (leobalter) [inline test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js]: this throws a TypeError but it does not reach the IsArray check. It returns the abrupt from IsConcatSpreadable step 2

The test and description are right, but the info can be adjusted.

Comment (leobalter) [inline test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js]: in anyway, we could have another test where the Proxy is revoked from handler.get trap which returns undefined.

File: test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Copyright (C) 2016 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-isconcatspreadable
+es6id: 22.1.3.1.1
```
Comment (leobalter) [inline test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js]: This is minor, but I think the ids should refer to [].concat in 22.1.3.1 / sec-array.prototype.concat

File: test/built-ins/JSON/parse/revived-proxy-revoked.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+assert.throws(TypeError, function() {
+  JSON.parse('[null, null]', function() {
+    callCount += 1;
+    this[1] = handle.proxy;
+    returnCount += 1;
+  });
+});
+
+assert.sameValue(callCount, 1, 'reviver is invoked exactly once');
```
Comment (leobalter) [inline test/built-ins/JSON/parse/revived-proxy-revoked.js]: the callCount seems unnecessary as we should not expect any normal completion before the returnCount increment

File: test/built-ins/JSON/parse/revived-proxy.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+arrayProxy.other = 0;
+
+injectProxy = function(name, val) {
+  if (name === 'other') {
+    visitedOther = true;
+  }
+  this[1] = objectProxy;
+  return val;
+};
```
Comment (leobalter) [inline test/built-ins/JSON/parse/revived-proxy.js]: wow!

wow

**Comment by leobalter** [comment]
Just a few parts mentioned above to fix.

File: test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js (hunk 3 above)
Comment (jugglinmike) [inline test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js]: Nice catch! The correct version of this test exposes a bug in V8

File: test/built-ins/JSON/parse/revived-proxy-revoked.js (hunk 5 above)
Comment (jugglinmike) [inline test/built-ins/JSON/parse/revived-proxy-revoked.js]: I suppose you're right. I think I was being a little too defensive here

File: test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js (hunk 4 above)
Comment (jugglinmike) [inline test/built-ins/Array/prototype/concat/is-concat-spreadable-proxy-revoked.js]: Following that pattern would make the IDs largely redundant with the information encoded in the test file names.

**Pull request merged**